### ISSUE TITLE
Implement backoff-delay retries for failed webhooks

### ISSUE THREAD

**Issue by dabeeeenster** [SolidStateGroup/bullet-train-api#1654]

**Comment by gagantrivedi** [comment]
Reopened by: https://github.com/Flagsmith/flagsmith/pull/2773

**Comment by tushar5526** [comment]
I don't think we can use the `backoff` library with tasks scheduled on task_processor (base on my understanding of it, we can not get the return value of scheduled tasks and catch exceptions raised by the tasks in task processor). We will have to write custom logic for backoff and jitter. 

We can register `_call_webhook` as a task and add a parameter to track the number of times it has been called. `call_webhook(webhook, data, try_count)`

The function will be then responsible for making recursive calls to itself whenever an exception is raised and until `try_count == 0`.

```python
@register_task_handler()
def _call_webhook(webhook, data, try_count):
    try:
       res = requests.post(str(webhook.url), data=json_data, headers=headers, timeout=10)
    except requests.exceptions.RequestException as exc:
       if try_count == 1:
           send_failure_email(webhook, data, webhook_type, res.status_code)
       else: 
           _call_webhook.delay(args=(webhoo, data, try_count - 1), backoff_time + jitter_time)
```

Error handling will be moved away to `_call_webhook` function.  

What are your thoughts on this approach? @gagantrivedi

**Comment by gagantrivedi** [comment]
> I don't think we can use the `backoff` library with tasks scheduled on task_processor (base on my understanding of it, we can not get the return value of scheduled tasks and catch exceptions raised by the tasks in task processor). We will have to write custom logic for backoff and jitter.
> 
> We can register `_call_webhook` as a task and add a parameter to track the number of times it has been called. `call_webhook(webhook, data, try_count)`
> 
> The function will be then responsible for making recursive calls to itself whenever an exception is raised and until `try_count == 0`.
> 
> ```python
> @register_task_handler()
> def _call_webhook(webhook, data, try_count):
>     try:
>        res = requests.post(str(webhook.url), data=json_data, headers=headers, timeout=10)
>     except requests.exceptions.RequestException as exc:
>        if try_count == 1:
>            send_failure_email(webhook, data, webhook_type, res.status_code)
>        else: 
>            _call_webhook.delay(args=(webhoo, data, try_count - 1), backoff_time + jitter_time)
> ```
> 
> Error handling will be moved away to `_call_webhook` function.
> 
> What are your thoughts on this approach? @gagantrivedi

Yeah, something along those lines
